mapie/time_series_regression.py

```diff
@@ -0,0 +1,312 @@
+from sklearn.utils import check_array
+from sklearn.utils.validation import check_is_fitted
+
+from ._compatibility import np_nanquantile
+from ._typing import ArrayLike, NDArray
+from .aggregation_functions import aggregate_all
+from .regression import MapieRegressor
+from .utils import (
+    check_alpha,
+    check_alpha_and_n_samples,
+)
+
+
+class MapieTimeSeriesRegressor(MapieRegressor):
+    """
+    Prediction intervals with out-of-fold residuals for time series.
+
+    This class implements the EnbPI strategy for estimating
+    prediction intervals on single-output time series. The only valid
+    ``method`` is 'enbpi'.
+
+    Actually, EnbPI only corresponds to ``MapieTimeSeriesRegressor`` if the
+    ``cv`` argument is of type ``BlockBootstrap``.
+
+    References
+    ----------
+    Chen Xu, and Yao Xie.
+    "Conformal prediction for dynamic time-series."
+    https://arxiv.org/abs/2010.09107
+    """
+
+    def __init__(
+        self,
+        estimator: Optional[RegressorMixin] = None,
+        method: str = "enbpi",
+        cv: Optional[Union[int, str, BaseCrossValidator]] = None,
+        n_jobs: Optional[int] = None,
+        agg_function: Optional[str] = "mean",
+        verbose: int = 0,
+    ) -> None:
+        super().__init__(estimator, method, cv, n_jobs, agg_function, verbose)
+        self.cv_need_agg_function.append("BlockBootstrap")
+        self.valid_methods_ = ["enbpi"]
+        self.plus_like_method.append("enbpi")
+
+    def _relative_conformity_scores(
+        self,
+        X: NDArray,
+        y: NDArray,
+    ) -> NDArray:
+        """
+        Compute the conformity scores on a data set.
+
+        Parameters
+        ----------
+            X : ArrayLike of shape (n_samples, n_features)
+            Input data.
+
+            y : ArrayLike of shape (n_samples,)
+                Input labels.
+
+        Returns
+        -------
+            The conformity scores corresponding to the input data set.
+        """
+        y_pred, _ = super().predict(X, alpha=0.5, ensemble=True)
+        return np.asarray(y) - np.asarray(y_pred)
+
+    def fit(
+        self,
+        X: ArrayLike,
+        y: ArrayLike,
+        sample_weight: Optional[ArrayLike] = None,
+    ) -> MapieTimeSeriesRegressor:
+        """
+        Compare to the method ``fit`` of ``MapieRegressor``, the ``fit`` method
+        of ``MapieTimeSeriesRegressor`` computes the ``conformity_scores_``
+        with relative values.
+
+        Returns
+        -------
+        MapieTimeSeriesRegressor
+            The model itself.
+        """
+        self = super().fit(X=X, y=y, sample_weight=sample_weight)
+        self.conformity_scores_ = self._relative_conformity_scores(X, y)
+        return self
+
+    def partial_fit(
+        self,
+        X: ArrayLike,
+        y: ArrayLike,
+    ) -> MapieTimeSeriesRegressor:
+        """
+        Update the ``conformity_scores_`` and ``k_`` attributes when new data
```

Comment:
Remove "and k_" now the "plus" method is removed ?

Reply:
Done